rfd/0085-improve-webapps-codebase.md

```diff
@@ -0,0 +1,163 @@
+as well as the final UI state.
+
+Tests should check for elements that exist for the specific state being tested, i.e. an error message when the data has 
+failed to load. Elements should be clicked on or events triggered in order to change the state of the component.
+
+By writing tests as described above, snapshot tests typically become redundant and therefore unnecessary. Besides the problem of skipping the business logic, snapshots consistently break for non-bugs, such as when styles are updated, and
+generally don't provide much if any security around behaviour or appearance. Further reading - 
+https://medium.com/@sapegin/whats-wrong-with-snapshot-tests-37fbe20dfe8e.
+
+### Changing how we export components
+
+Default exports should be avoided except when absolutely necessary (for example, with 
+`React.Lazy`), as they degrade the developer experience
+for no discernible benefit. For more fully fleshed out reasoning, refer to the articles below:
+- https://blog.neufund.org/why-we-have-banned-default-exports-and-you-should-do-the-same-d51fdc2cf2ad
+- https://rajeshnaroth.medium.com/avoid-es6-default-exports-a24142978a7a
+- https://blog.piotrnalepa.pl/2020/06/26/default-exports-vs-named-exports/
+- https://ilikekillnerds.com/2019/08/default-exports-bad/
+
+When needing to export something as default, we should export it like so:
+
+```typescript jsx
+function SomeComponent() {
+  return (
+    <div>
+      hello!
+    </div>
+  );
+}
+
+export { SomeComponent as default };
+```
+
+This provides a clearer, more explicit export.
+
+### Code Coverage
+
+Our current code coverage is:
+
+```
+=============================== Coverage summary ===============================
+Statements   : 36.12% ( 3843/10639 )
+Branches     : 39.44% ( 1484/3762 )
+Functions    : 32.36% ( 1376/4251 )
+Lines        : 36.31% ( 3759/10350 )
+================================================================================
+```
+
+We should add code coverage to CI in order to track progress. One tool we can use for this is 
+[Codecov](https://about.codecov.io/), which provide a free plan for open source tools. This would provide a useful UI
+for inspecting coverage issues. It can also annotate pull requests with coverage issues.
+
+We should then implement Codecov's status check on pull requests. This check should fail if a certain % of the code 
+touched by the pull request has not got coverage. This means that developers won't be punished for the lack of coverage
+in the existing codebase, as it'll only check the coverage percentage of the changes in the pull request.
+
+### How to implement these changes
+
+From this RFD, all new code to the webapps repo should follow this convention. It would take a while to go back through
+the existing codebase and make these changes, as it would involve rewriting the tests that already exist. An effort to
+improve code coverage would be beneficial, but not a priority.
```

Comment:
I love lines 1-158 and less-love lines 159-163.

I understand it might not be a top priority right now, but I would love to see us take some actionable steps beyond "awareness." This doc becomes a bit of tribal knowledge, and any person ramping in our codebase will see and adhere to patterns we don't want to exist. And in turn, we'll ultimately point them in the direction of said RFD and say "read."

Can we carve out even minor GH issues that can be worked on in short cycles (an afternoon, a day, or two) between priority work? Chipping away intentionally vs. refactor as we go would make more progress. I feel there are parts of our codebase that we rarely change these days.

Reply:
I'm all for going back and chipping away with refactoring (I was going to do this after the RFD anyhow), I just didn't know if it was wise to make it a requirement - if we get buy in that we can allocate resources to cleaning up, that would be even better.

Reply:
I've reworded this